File: equity-volume-breadth/qa/plain_language/browser_qa_runner.py
#!/usr/bin/env python3
"""Browser QA for the generated investor report."""

import hashlib
import json
from pathlib import Path

from playwright.sync_api import sync_playwright


BASE = Path("/Users/bobo/ZCodeProject/equity-volume-breadth")
HTML = BASE / "index.html"
OUT = BASE / "qa" / "plain_language" / "browser_qa.json"
SHOTS = BASE / "screenshots"


def main():
    SHOTS.mkdir(parents=True, exist_ok=True)
    results = {
        "html_sha256": hashlib.sha256(HTML.read_bytes()).hexdigest(),
        "viewports": {},
        "console_errors": [],
        "overflow_issues": [],
    }

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        for name, width, height in (
            ("desktop", 1440, 1100),
            ("mobile-390", 390, 844),
            ("small-320", 320, 800),
        ):
            page = browser.new_page(viewport={"width": width, "height": height})
            errors = []
            page.on("console", lambda msg: errors.append(msg.text) if msg.type == "error" else None)
            page.on("pageerror", lambda error: errors.append(str(error)))
            page.goto(HTML.as_uri(), wait_until="networkidle")
            page.wait_for_timeout(1500)

            state = page.evaluate(
                """() => {
                    const canvases = [...document.querySelectorAll('canvas')];
                    const ids = [...document.querySelectorAll('[id]')].map(e => e.id);
                    const navTargets = [...document.querySelectorAll('#mm a[href^="#"]')]
                      .map(a => a.getAttribute('href').slice(1));
                    return {
                      canvas_count: canvases.length,
                      canvas_rendered: canvases.filter(c => c.width > 0 && c.height > 0).length,
                      canvas_details: canvases.map(c => ({id:c.id,w:c.offsetWidth,h:c.offsetHeight})),
                      scrollWidth: document.documentElement.scrollWidth,
                      innerWidth: innerWidth,
                      duplicate_ids: [...new Set(ids.filter((id, i) => ids.indexOf(id) !== i))],
                      missing_nav_targets: navTargets.filter(id => !document.getElementById(id)),
                      conclusion_present: !!document.getElementById('ketluan'),
                    };
                }"""
            )
            state["overflow"] = state["scrollWidth"] > state["innerWidth"]
            state["console_errors"] = errors
            results["viewports"][name] = state
            if errors:
                results["console_errors"].append({"viewport": name, "errors": errors})
            if state["overflow"]:
                results["overflow_issues"].append({"viewport": name})

            page.screenshot(path=str(SHOTS / f"{name}-hero.png"))
            page.screenshot(path=str(SHOTS / f"{name}-full.png"), full_page=True)
            if name == "desktop":
                page.locator("#ketluan").screenshot(path=str(SHOTS / "desktop-conclusion.png"))
                page.locator("#oos").screenshot(path=str(SHOTS / "desktop-oos.png"))
            page.close()
        browser.close()

    states = list(results["viewports"].values())
    results["gate_summary"] = {
        "charts_rendered_ok": all(s["canvas_count"] == 3 and s["canvas_rendered"] == 3 for s in states),
        "zero_console_errors": not results["console_errors"],
        "no_horizontal_overflow": not results["overflow_issues"],
        "no_duplicate_ids": all(not s["duplicate_ids"] for s in states),
        "nav_targets_resolve": all(not s["missing_nav_targets"] for s in states),
        "conclusion_present": all(s["conclusion_present"] for s in states),
    }
    results["gate_summary"]["overall_pass"] = all(results["gate_summary"].values())
    OUT.write_text(json.dumps(results, ensure_ascii=False, indent=2))
    print(json.dumps(results["gate_summary"], ensure_ascii=False))
    if not results["gate_summary"]["overall_pass"]:
        raise SystemExit(1)


if __name__ == "__main__":
    main()

File: equity-volume-breadth/qa/plain_language/browser_qa_runner.js
const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

(async () => {
  const browser = await chromium.launch();
  const url = 'file:///Users/bobo/ZCodeProject/equity-volume-breadth/index.html';
  const shotDir = '/Users/bobo/ZCodeProject/equity-volume-breadth/screenshots';
  const viewports = [
    { name: 'desktop', w: 1440, h: 1100 },
    { name: 'mobile-390', w: 390, h: 844 },
    { name: 'small-320', w: 320, h: 800 },
  ];
  const results = { viewports: {}, charts_rendered: 0, console_errors: [], overflow_issues: [] };

  for (const vp of viewports) {
    const ctx = await browser.newContext({ viewport: { width: vp.w, height: vp.h } });
    const page = await ctx.newPage();
    const errors = [];
    page.on('console', msg => { if (msg.type() === 'error') errors.push(msg.text()); });
    page.on('pageerror', err => errors.push(String(err)));

    await page.goto(url, { waitUntil: 'networkidle' });
    await page.waitForTimeout(2000);

    // Check canvas rendered (non-zero size)
    const canvasInfo = await page.evaluate(() => {
      const canvases = document.querySelectorAll('canvas');
      return Array.from(canvases).map(c => ({
        id: c.id, w: c.offsetWidth, h: c.offsetHeight,
        hasContent: c.width > 0 && c.height > 0
      }));
    });

    // Check horizontal overflow
    const overflow = await page.evaluate(() => ({
      scrollWidth: document.documentElement.scrollWidth,
      innerWidth: window.innerWidth,
    }));
    const overflows = overflow.scrollWidth > overflow.innerWidth;

    // Screenshot (hero + full)
    await page.screenshot({ path: path.join(shotDir, `${vp.name}-hero.png`) });
    await page.screenshot({ path: path.join(shotDir, `${vp.name}-full.png`), fullPage: true });

    // For desktop also screenshot the OOS section
    if (vp.name === 'desktop') {
      const oosEl = await page.$('#oos-giai-thich');
      if (oosEl) await oosEl.screenshot({ path: path.join(shotDir, 'desktop-oos.png') });
      const mainEl = await page.$('#dien-truoc');
      if (mainEl) await mainEl.screenshot({ path: path.join(shotDir, 'desktop-results.png') });
      results.charts_rendered = canvasInfo.filter(c => c.hasContent).length;
    }

    results.viewports[vp.name] = {
      canvas_count: canvasInfo.length,
      canvas_rendered: canvasInfo.filter(c => c.hasContent).length,
      canvas_details: canvasInfo,
      console_errors: errors,
      scrollWidth: overflow.scrollWidth,
      innerWidth: overflow.innerWidth,
      overflow: overflows,
    };
    if (errors.length) results.console_errors.push({ viewport: vp.name, errors });
    if (overflows) results.overflow_issues.push({ viewport: vp.name, scrollWidth: overflow.scrollWidth, innerWidth: overflow.innerWidth });

    await ctx.close();
  }

  await browser.close();

  // Gate checks
  const allChartsRender = Object.values(results.viewports).every(v => v.canvas_rendered === 3);
  const noConsoleErrors = results.console_errors.length === 0;
  const noOverflow = results.overflow_issues.length === 0;

  results.gate_summary = {
    charts_rendered_ok: allChartsRender,
    zero_console_errors: noConsoleErrors,
    no_horizontal_overflow: noOverflow,
    overall_pass: allChartsRender && noConsoleErrors && noOverflow,
  };

  fs.writeFileSync(
    '/Users/bobo/ZCodeProject/equity-volume-breadth/qa/plain_language/browser_qa.json',
    JSON.stringify(results, null, 2)
  );
  console.log('GATE:', results.gate_summary.overall_pass ? 'PASS' : 'FAIL');
  console.log('charts:', results.charts_rendered, '/ 4');
  console.log('console errors:', results.console_errors.length);
  console.log('overflow issues:', results.overflow_issues.length);
  process.exit(results.gate_summary.overall_pass ? 0 : 1);
})();
